src/pyhf_combine_converter/pyhf_convert_to_datacard.py

```diff
@@ -299,232 +309,204 @@ def addMods():  ##add systematics to DC
-                            + " "
-                            + str(0)
-                            + " "
-                            + str(2)
-                            + "\n"
-                        )
-        f.close()
+            if (
+                channel in data_card.binParFlags.keys()
+                and data_card.binParFlags[channel] == True
+            ):
+                # double check to be safe
+                shapesys = False
+                staterror = False
+                for sample in spec["channels"][idxc]["samples"]:
+                    mod_types = [mod["type"] for mod in sample["modifiers"]]
+                    if "shapesys" in mod_types:
+                        shapesys = True
+                    elif "staterror" in mod_types:
+                        staterror = True
+
+                if shapesys:
+                    f.write(f"{channel} autoMCStats 100000 0 2" + "\n")
+                if staterror:
+                    f.write(f"{channel} autoMCStats 0 0 2" + "\n")
+
+
+def main():
+    parser = OptionParser()  # add command line args
+    parser.add_option(
+        "-O",
+        "--out-datacard",
+        dest="outdatacard",
+        default="converted_datacard.txt",
+        help="desired name of datacard file",
+    )
+    parser.add_option(
+        "-s",
+        "--shape-file",
+        dest="shapefile",
+        default="shapes.root",
+        help="desired name of shapes file",
+    )
+    options, args = parser.parse_args()
 
-if __name__ == "__main__":
     with open(args[0]) as serialized:
         spec = json.load(serialized)
     workspace = pyhf.Workspace(spec)
     model = workspace.model()
 
-
     channels = model.config.channels
     channel_bins = model.config.channel_nbins
     samples = model.config.samples
     modifiers = model.config.modifiers
     lumi = model.config.parameters
 
-    ##generate list of mods without normfactors, shapesys, staterror
-    systs = []
-    for mod in modifiers:
-        if "normfactor" not in mod:
-            if "shapesys" not in mod:
-                if "staterror" not in mod:
-                    systs.append(mod)
-
-
-    DC = Datacard() 
+    # generate list of mods without normfactors, shapesys, staterror
+    systs = [
+        mod
+        for mod in modifiers
+        if "normfactor" not in mod and "shapesys" not in mod and "staterror" not in mod
+    ]
 
+    data_card = Datacard()
 
     file = uproot.recreate(options.shapefile)
-    addChannels()
-    addSamples()
-    addMods()
-    addSignal()
-    addRateParams()
+    addChannels(file, spec, data_card, channel_bins)
+    addSamples(file, spec, data_card, channel_bins, samples, options)
+    addMods(file, spec, data_card, channel_bins, systs)
+    addSignal(spec, data_card, channels, modifiers)
+    addRateParams(spec, data_card, channels, modifiers)
     file.close()
-    writeDataCard(options.outdatacard)
+    write_data_card(spec, data_card, channels, options.outdatacard)
```

Comment:
This is just passing in the necessary data structures into the functions, but as I've revised the API a bit here please make sure that the tests that you've run so far work as expected before any of this gets merged. :+1: